templates/boltzmann.js

```diff
@@ -575,7 +579,105 @@ function applyHeaders (headers = {}) {
   }
 }
 
-const applyXFO = (mode) => applyXFO.bind({ 'x-frame-options': mode })
+const applyXFO = (mode) => applyHeaders({ 'x-frame-options': mode })
+
+// {% if csrf %}
+// csrf protection middleware
+function signCookie(value, secret) {
+  return `${value}.${crypto.createHmac('sha256', secret).update(value).digest('base64')}`
+}
+
+function checkCookieSignature(input, secret) {
+  if (!input) {
+    return false
+  }
+  const [ message, signature ] = input.split('.', 2)
+  const valid = signCookie(message, secret)
+  return crypto.timingSafeEqual(Buffer.from(input), Buffer.from(valid)) ? message : false
+}
+
+const READ_METHODS = new Set(['GET', 'HEAD', 'OPTIONS'])
+
+function applyCSRF ({
+  cookieSecret = process.env.COOKIE_SECRET,
+  csrfCookie = '_csrf',
+  param = '_csrf',
+  header = 'csrf-token'
+} = {}) {
+
+  if (!cookieSecret) {
+    throw new Error('You cannot use CSRF middleware without providing a secret for signing cookies')
+  }
+
+  return next => {
+    const tokens = new CsrfTokens()
+    return async function csrf (context) {
+
+      function generateNewSecretCookie () {
+        const newSecret = tokens.secretSync()
+        const signed = signCookie(newSecret, cookieSecret)
+        context.cookie.set(csrfCookie, signed)
+        return newSecret
+      }
+
+      function fetchSecretFromCookie () {
+        const candidate = context.cookie.get(csrfCookie)
+        if (!candidate) {
+          return undefined
+        }
+        return checkCookieSignature(candidate.value, cookieSecret)
+      }
+
+      async function findToken (context) {
+        const body = await context.body
+        return (body && body[param]) || context.headers[header]
+      }
+
+      // Handlers can call this to get a token to use on relevant requests.
+      // It creates a token-generating secret for the user if they don't have one
+      // already, and makes a new token.
+      function csrfToken () {
+        const freshSecret = fetchSecretFromCookie()
+
+        // We might be coming through here more than once.
+        // Re-use the token if we can, but generate a new one if the secret in the cookie changed.
+        if (token && freshSecret === secret) {
+          return token
+        }
+
+        if (!freshSecret) {
+          secret = generateNewSecretCookie() // changes value in the closure
+        }
+
+        token = tokens.create(secret) // changes value in the closure
+        return token
+      }
+
+      // set up context for handler, with intentional hoist
+      context.csrfToken = csrfToken
```

Comment:
We should expose a method for handlers to use to force a CSRF token refresh – for example, after login or logout.